### PR TITLE
MOTECH-2998 Added time window to Task configuration

### PR DESCRIPTION
**Pull request by mjazdzdewski**

### CODE REVIEW

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

File: modules/tasks/tasks/src/main/resources/webapp/js/directives.js [hunk 1]
```diff
@@ -1045,6 +1045,53 @@
...
+                            timeFormat: 'HH:mm z',
+                            onSelect: function (time) {
+                                ngModelCtrl.$setViewValue(time);
+                                scope.$apply(function() {
+                                    scope.$parent.startTime = time.toString();
+                                });
+                            }
+                        });
+                    });
+
```
Comment (pkornowski) [inline modules/tasks/tasks/src/main/resources/webapp/js/directives.js]: Please, remove empty line.

File: modules/tasks/tasks/src/main/resources/webapp/js/directives.js [hunk 2]
```diff
@@ -1045,6 +1045,53 @@
...
+                }
+            };
+        });
+
+
```
Comment (pkornowski) [inline modules/tasks/tasks/src/main/resources/webapp/js/directives.js]: Remove this line.

**Comment by pkornowski** [comment]
@kmadej Looks good to me.

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/domain/mds/task/Task.java [hunk 3]
```diff
@@ -78,6 +79,15 @@
     @Field
     private int retryIntervalInMilliseconds;
 
+    @Field
+    private boolean useTimeWindow;
+
+    @Field
+    private Time startTime;
```
Comment (kmadej) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/domain/mds/task/Task.java]: use DateTime

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/domain/mds/task/Task.java [hunk 4]
```diff
@@ -118,6 +128,9 @@ public Task(String name, TaskTriggerInformation trigger, List<TaskActionInformat
         this.failuresInRow = 0;
         this.numberOfRetries = DEFAULT_NUMBER_FOR_TASK_RETRIES;
         this.retryIntervalInMilliseconds = DEFAULT_TIME_FOR_RETRY_INTERVAL;
+        this.useTimeWindow = false;
+        this.startTime = null;
+        this.endTime = null;
```
Comment (kmadej) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/domain/mds/task/Task.java]: this is redundant

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/impl/TaskServiceImpl.java [hunk 5]
```diff
@@ -192,7 +195,16 @@ public ActionEvent getActionEventFor(TaskActionInformation taskActionInformation
...
+                    String isNotUsingTimeWindow = "useTimeWindow != true";
+                    String now = (new Time(LocalTime.now(DateTimeZone.UTC))).toString();
+                    String isAfterStartTime = String.format("startTime < \"%s\"", now);
+                    String isBeforeEndTime = String.format("endTime > \"%s\"", now);
+                    String isEndTimeNextDay = "endTime < startTime";
+                    String isBetweenStartTimeAndMidnight = String.format("(%s) && (\"24:00\" > \"%s\")", isAfterStartTime, now);
+                    String isBetweenMidnightAndEndTime = String.format("(%s) && (\"00:00\" < \"%s\")", isBeforeEndTime, now);
+                    String isInTimeWindow = String.format("(%s) || ((%s) && (%s)) || ((%s) && ((%s) || (%s)))",
+                            isNotUsingTimeWindow, isAfterStartTime, isBeforeEndTime, isEndTimeNextDay, isBetweenStartTimeAndMidnight, isBetweenMidnightAndEndTime);
+                    String filter = String.format("(%s) && (%s) && (%s)", isTaskActive, byTriggerSubject, isInTimeWindow);
```
Comment (kmadej) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/impl/TaskServiceImpl.java]: it should be done simplier, search for sql queries with dates

File: modules/tasks/tasks/src/main/resources/webapp/js/controllers.js [hunk 6]
```diff
@@ -932,6 +966,42 @@
...
+            hoursStr = hours.toString();
+            minutesStr = minutes.toString();
+            if(hours < 10) {
+                hoursStr = "0" + hoursStr;
+            }
+            if(minutes < 10) {
+                minutesStr = "0" + minutesStr;
+            }
+            formatted = hoursStr + (":") + minutesStr;
+            return formatted;
```
Comment (kmadej) [inline modules/tasks/tasks/src/main/resources/webapp/js/controllers.js]: it looks a bit complicated, I think you should use sth external used in motech already

**Comment by kmadej** [comment]
I have reviewed this once again, and the changes I suggested might take too much time, so the only thing to do is moving time formatting to UTC from frontend to backend

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/impl/TaskServiceImpl.java [hunk 7]
```diff
@@ -192,7 +195,16 @@ public ActionEvent getActionEventFor(TaskActionInformation taskActionInformation
...
```
Comment (kmadej) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/impl/TaskServiceImpl.java]: get list of task from sql and then parse this list, it should be simplier than doing this using raw sql

File: modules/tasks/tasks/src/main/resources/webapp/js/controllers.js [hunk 8]
```diff
@@ -435,21 +442,48 @@
...
+                    }
+                    if (regexElements[3] > 1400) {
+                        return false;
+                    }
+                } else {
+                    return false;
+                }
+            }
+            return true;
+        };
```
Comment (kmadej) [inline modules/tasks/tasks/src/main/resources/webapp/js/controllers.js]: is it required ?

File: platform/commons-date/src/main/java/org/motechproject/commons/date/model/Time.java [hunk 9]
```diff
@@ -50,17 +55,24 @@ public Time(LocalTime localTime) {
      * @throws IllegalArgumentException if {@code timeStr} doesn't match "HH:MM" pattern
      */
     public Time(String timeStr) {
-        String[] tokens = timeStr.split(":");
-
-        if (tokens.length < TIME_TOKEN_MIN_LENGTH || tokens.length > TIME_TOKEN_MAX_LENGTH) {
-            throw new IllegalArgumentException("Invalid time string: " + timeStr);
-        }
+        if (timeStr.length() == 11) {
```
Comment (kmadej) [inline platform/commons-date/src/main/java/org/motechproject/commons/date/model/Time.java]: move this to separate method 'isTimeContainsZone

**Pull request merged**